tls/s2n_handshake.c

```diff
@@ -329,3 +329,20 @@ struct s2n_cert_chain_and_key *s2n_get_compatible_cert_chain_and_key(struct s2n_
         return conn->config->default_certs_by_type.certs[cert_type];
     }
 }
+
+/* This method will work when testing S2N, and for the EndOfEarlyData message.
+ *
+ * However, it will NOT work for arbitrary message types when potentially receiving records
+ * that contain multiple messages, like when talking to a non-S2N TLS implementation. If the "end_message"
+ * is not the first message in a multi-message record, negotiation will not stop.
+ * (This is not an issue for EndOfEarlyData because encryption and message order requirements force
+ * EndOfEarlyData to always be the first and only handshake message in its handshake record)
+ */
+int s2n_negotiate_until_message(struct s2n_connection *conn, s2n_blocked_status *blocked, message_type_t end_message)
```

Comment:
Why is this in the handshake file if it is a testing method?

Reply:
nvm i see the non-test use-case

Reply:
It's not just a testing method. It's used by a testing method (s2n_negotiate_test_server_and_client_until_message) but the main usecase is for early data handshakes, where we need to stop negotiation at a certain point until we finish sending all the early data we intend to send.